subsys/bluetooth/mesh/adv_ext.c

```diff
@@ -506,6 +506,26 @@ int bt_mesh_adv_enable(void)
 	return 0;
 }
 
+int bt_mesh_adv_disable(void)
```

Comment:
Discussed offline. All adv attempts will fail here: https://github.com/zephyrproject-rtos/zephyr/blob/4fe2605160fd535980924a7ea09080c5984fe896/subsys/bluetooth/mesh/adv_ext.c#L147-L150

But do you know what happens with already scheduled advertising? Is `adv_sent` callback is called? If so, won't such advertisement be lost here: https://github.com/zephyrproject-rtos/zephyr/blob/4fe2605160fd535980924a7ea09080c5984fe896/subsys/bluetooth/mesh/adv_ext.c#L453-L458

I tend to think that we need the following test:
- Fill up the advertiser pool.
- When the first advertisement is started (`struct bt_mesh_send_cb.start` is called), call `bt_mesh_suspend()`.
- Expect that `struct bt_mesh_send_cb.end` is called for all advertisements.
- While the stack is suspended, try to schedule one more advertisement. 
- Expect that `struct bt_mesh_send_cb.start` is called with error (or what behavior do we want here?).
- Call `be_mesh_resume()` and fill up the adv pool again.
- Expect that all advertisements are sent successfully.

Feel free to create a separate task for this test as it was not planned, but I think this test will guarantee that advertiser behavior works as we want. It will also guarantee that the adv pool is emptied correctly and can be reused later. The test with publication IMHO seems very high level and don't give enough confident that suspend/resume behavior works correctly.

Reply:
Regarding the initial comment, we arrived at the conclusion that there was no need to unref buffers. But should still `bt_mesh_adv_send_start()` be called? If so, with what error code? Is this sufficient for the "clean-up" part, or is there more that should be done?

The already scheduled advertising is cancelled, with [this line](https://github.com/zephyrproject-rtos/zephyr/blob/e13912d92323c6ba94665c898408a18ef4cd7f3c/subsys/bluetooth/mesh/adv_ext.c#L514). 

I will create a task for the final part, thanks for the input!

Reply:
> Regarding the initial comment, we arrived at the conclusion that there was no need to unref buffers. But should still `bt_mesh_adv_send_start()` be called? If so, with what error code? Is this sufficient for the "clean-up" part, or is there more that should be done?

True, it is not needed as `bt_mesh_adv_send_start()` will be called here with an error:
https://github.com/zephyrproject-rtos/zephyr/blob/e13912d92323c6ba94665c898408a18ef4cd7f3c/subsys/bluetooth/mesh/adv_ext.c#L231

> The already scheduled advertising is cancelled, with [this line](https://github.com/zephyrproject-rtos/zephyr/blob/e13912d92323c6ba94665c898408a18ef4cd7f3c/subsys/bluetooth/mesh/adv_ext.c#L514).

Sorry, by scheduled I meant pushed to the host by this call: https://github.com/zephyrproject-rtos/zephyr/blob/e13912d92323c6ba94665c898408a18ef4cd7f3c/subsys/bluetooth/mesh/adv_ext.c#L178
`adv->buf` will contain reference to that buffer (https://github.com/zephyrproject-rtos/zephyr/blob/e13912d92323c6ba94665c898408a18ef4cd7f3c/subsys/bluetooth/mesh/adv_ext.c#L228), which is then unreferenced here: https://github.com/zephyrproject-rtos/zephyr/blob/e13912d92323c6ba94665c898408a18ef4cd7f3c/subsys/bluetooth/mesh/adv_ext.c#L267-L270
So the question is will it work if adv->instance is NULL (https://github.com/zephyrproject-rtos/zephyr/blob/e13912d92323c6ba94665c898408a18ef4cd7f3c/subsys/bluetooth/mesh/adv_ext.c#L527)? Or is `adv_send` callback is called in `bt_le_ext_adv_stop` (https://github.com/zephyrproject-rtos/zephyr/blob/e13912d92323c6ba94665c898408a18ef4cd7f3c/subsys/bluetooth/mesh/adv_ext.c#L516) and then there is no issue?

Reply:
Discussed offline.